openverse_catalog/dags/providers/provider_api_scripts/freesound.py

```diff
@@ -0,0 +1,266 @@
+    license_name="",
+    page_number=1,
+    default_query_params=None,
+):
+    default_query_params = default_query_params or DEFAULT_QUERY_PARAMS
+    return {
+        **default_query_params,
+        "page": str(page_number),
+        "license": license_name,
+    }
+
+
+def _get_items(license_name):
+    item_count = 0
+    page_number = 1
+    should_continue = True
+    while should_continue:
+        query_param = _get_query_params(
+            license_name=license_name, page_number=page_number
+        )
+        batch_data = _get_batch_json(query_param=query_param)
+        if batch_data:
+            item_count = _process_item_batch(batch_data)
+            page_number += 1
+        else:
+            should_continue = False
+    return item_count
+
+
+def _get_batch_json(endpoint=ENDPOINT, headers=None, retries=RETRIES, query_param=None):
+    if headers is None:
+        headers = HEADERS
+    response_json = delayed_requester.get_response_json(
+        endpoint, retries, query_param, headers=headers
+    )
+    if response_json is None:
+        return None
+    else:
+        results = response_json.get("results")
+        return results
+
+
+def _process_item_batch(items_batch):
+    for item in items_batch:
+        item_meta_data = _extract_audio_data(item)
+        if item_meta_data is None:
+            continue
+        audio_store.add_item(**item_meta_data)
+    return audio_store.total_items
+
+
+def _extract_audio_data(media_data):
+    """Extracts meta data about the audio file
+    Freesound does not have audio thumbnails"""
+    foreign_landing_url = media_data.get("url")
+    if not foreign_landing_url:
+        return None
+
+    foreign_identifier = media_data.get("id")
+    if not foreign_identifier:
+        return None
+
+    item_license = _get_license(media_data)
+    if item_license is None:
+        return None
+
+    # We use the mp3-hq url as `audio_url` for playing on the frontend
+    # download link (that requires authentication) for download button
+    # media_data['url'] as the foreign_landing_url
+    main_audio, alt_files = _get_audio_files(media_data)
+    if main_audio is None:
+        return None
+
+    creator, creator_url = _get_creator_data(media_data)
+    duration = int(media_data.get("duration") * 1000)
+    set_foreign_id, audio_set, set_url = _get_audio_set_info(media_data)
+    return {
+        "title": media_data.get("name"),
+        "creator": creator,
+        "creator_url": creator_url,
+        "foreign_identifier": foreign_identifier,
+        "foreign_landing_url": foreign_landing_url,
+        "duration": duration,
+        "license_info": item_license,
+        "meta_data": _get_metadata(media_data),
+        "raw_tags": media_data.get("tags"),
+        "set_foreign_id": set_foreign_id,
+        "audio_set": audio_set,
+        "set_url": set_url,
+        "alt_files": alt_files,
+        "category": FREESOUND_CATEGORY,
+        # audio_url, filetype, bit_rate
```

Comment:
Seems like this is still a WIP (?) I believe these fields are derived from `main_audio` here.

Reply:
Line 164 replaces the main_audio object with the hq mp3 preview and its properties. Or, in an unlikely case previews don't exist, it doesn't.

Reply:
I didn't notice this a was a comment for the next line, with the unpacking operator, my bad, it's all good 👌